programs/uxd/src/instructions/mercurial/collect_profit_of_mercurial_vault_depository.rs

```diff
@@ -0,0 +1,298 @@
+    fn calculate_lp_token_amount_to_match_collectable_profits_value(
+        &self,
+        target_value: u64,
+    ) -> Result<u64> {
+        let current_time = u64::try_from(Clock::get()?.unix_timestamp)
+            .ok()
+            .ok_or_else(|| error!(UxdError::MathError))?;
+
+        // Because it's u64 type, we will never withdraw too much due to precision loss, but withdraw less.
+        // We withdraw less interests and fee due to precision loss and that's ok
+        self.mercurial_vault
+            .get_unmint_amount(
+                current_time,
+                target_value,
+                self.mercurial_vault_lp_mint.supply,
+            )
+            .ok_or_else(|| error!(UxdError::MathError))
+    }
+
+    pub fn calculate_collectable_profits_value(&self) -> Result<u64> {
+        let owned_lp_tokens_value = I80F48::checked_from_num(
+            mercurial_utils::calculate_lp_tokens_value::calculate_lp_tokens_value(
+                &self.mercurial_vault,
+                self.mercurial_vault_lp_mint.supply,
+                self.depository_lp_token_vault.amount,
+            )?,
+        )
+        .ok_or_else(|| error!(UxdError::MathError))?;
+
+        let collectable_profits_amount = owned_lp_tokens_value
+            .checked_sub(
+                I80F48::checked_from_num(
+                    self.depository.load()?.redeemable_amount_under_management,
+                )
+                .ok_or_else(|| error!(UxdError::MathError))?,
+            )
+            .ok_or_else(|| error!(UxdError::MathError))?;
+
+        collectable_profits_amount
+            .checked_to_num()
+            .ok_or_else(|| error!(UxdError::MathError))
+    }
+
+    // Check that the collateral amount received by the user matches the collateral amount we wanted the user to receive:
+    //     redeemable amount - fees - lp token calculation precision loss - withdraw precision loss
+    fn check_redeemed_collateral_amount_to_match_target(
+        redeemed_collateral_amount: u64,
+        target: u64,
+        possible_lp_token_precision_loss_collateral_value: u64,
+    ) -> Result<()> {
+        // Lp token precision loss + withdraw collateral precision loss
+        let maximum_allowed_precision_loss = possible_lp_token_precision_loss_collateral_value
+            .checked_add(1)
+            .ok_or_else(|| error!(UxdError::MathError))?;
+
+        let target_minimal_allowed_value = target
+            .checked_sub(maximum_allowed_precision_loss)
+            .ok_or_else(|| error!(UxdError::MathError))?;
+
+        require!(
+            (target_minimal_allowed_value..(target + 1)).contains(&redeemed_collateral_amount),
+            UxdError::SlippageReached,
+        );
+
+        Ok(())
+    }
+}
+
+// Validate
+impl<'info> CollectProfitOfMercurialVaultDepository<'info> {
+    pub(crate) fn validate(&self) -> Result<()> {
+        validate_is_program_frozen(self.controller.load()?)?;
+        require!(
```

Comment:
I was thinking actually this is not required since we already check that the account is a valid token account in the annotation:
https://github.com/UXDProtocol/uxd-program/pull/214/files#diff-b05c2c79a3b28a2632affdf9b4c6868e66fed2d33da06ce9b6bc7a5523119a73R50

Reply:
yes, removable